test/libsolidity/semanticTests/array/create_memory_array.sol

```diff
@@ -18,6 +18,6 @@ contract C {
 }
 // ----
 // f() -> "A", 8, 4, "B"
-// gas irOptimized: 125822
+// gas irOptimized: 136664
```

Comment:
In this one the default sequence seems to win mostly due to efficient use of `FunctionSpecializer` (`F`).

The heaviest parts of the code are the two allocations of 2D arrays, which turn into loops in Yul. For example the first one looks like this with the default sequence:
```yul
                        for { } lt(i, 9600) { i := add(i, _5) }
                        {
                            let memPtr_1 := allocate_memory()
                            calldatacopy(memPtr_1, calldatasize(), _4)
                            mstore(add(add(memPtr, i), _5), memPtr_1)
                        }
```
and like this with the new sequence:
```yul
                        for { } lt(i, _3) { i := add(i, 32) }
                        {
                            let size_2 := 0
                            let _4 := 0
                            _4 := 0
                            size_2 := 64
                            let memPtr_2 := allocate_memory(size_2)
                            let size_3 := 0
                            let _5 := 0
                            _5 := 0
                            size_3 := size_2
                            calldatacopy(memPtr_2, calldatasize(), size_2)
                            mstore(add(add(memPtr_1, i), 32), memPtr_2)
                        }
```

Surprisingly, the bit that accounts for most of the cost increase is the parameter passed into `allocate_memory()`. The default sequence inlines the uses of the function outside of the loop and then can specialize it for the loop, making it significantly cheaper. Other differences are much less significant.

I managed to replicate this behavior with an extra `TFviu` at the end of the new sequence (the `v` is important, since without it the function gets completely inlined). The downside of this is that it significantly hurts some other examples. E.g. the bytecode size in `ramanujan_pi.sol` blows up by ~20%. The default sequence seems to be able to avoid it, probably due to repetition of other sequence parts so maybe adding some of them at the end could help. I also noticed that I can move the new segment a bit earlier in the sequence by prepending some steps to it (e.g. `ciTFviu`) and this makes the other examples degrade less.

Another change that improves this test case a little is adding `Tr` to the cleanup sequence (i.e. `fDnTOcTrUmu`).

Overall, this one could possibly be improved with more experimentation, but it's much harder than the other tweaks.

Reply:
Actually, regression for `ramanujan_pi.sol` is not as bad as it seemed at first, because I was comparing with the original results for the new sequence. With the above tweaks and compared with the default sequence it's only 8%. Still, it was -6% before so by doing that we just trade one regression for another.

Reply:
### Comparison with unmodified `the-good-parts-mk2` on `cancun`
#### Tweak
```diff
+       "ciTFviu"
        "[scCTUt] TOntnfDIul"          // Perform structural simplification
        "gvif"                         // Run full inliner

        "jmul[jul] VcTOcul jmul";      // Make source short and pretty

-   static char constexpr DefaultYulOptimiserCleanupSteps[] = "fDnTOcmu";
+   static char constexpr DefaultYulOptimiserCleanupSteps[] = "fDnTOcTrmu";
```

#### External tests
[`seqbench-sequence-the-good-parts-mk2-tweaks-ciTFviu-Tr`](https://app.circleci.com/pipelines/github/ethereum/solidity/33673/workflows/e24c1af1-8d6a-4a79-aa96-a7797496135a) vs [`seqbench-sequence-the-good-parts-mk2-cancun`](https://app.circleci.com/pipelines/github/ethereum/solidity/33655/workflows/f901ee55-094e-44b1-860c-44cab6b675bb)

##### `ir-optimize-evm+yul`
|          project |  bytecode_size | deployment_gas |     method_gas |
|:----------------:|---------------:|---------------:|---------------:|
|            brink | **`+1.24% ❌`** |                |                |
|           colony | **`-0.01% ✅`** |                |                |
|        elementfi | **`+2.22% ❌`** |                |                |
|              ens | **`+0.29% ❌`** | **`-1.04% ✅`** | **`-0.04% ✅`** |
|            euler | **`-0.13% ✅`** |                |                |
|              gp2 | **`+0.25% ❌`** |                |                |
|  perpetual-pools | **`-2.11% ✅`** | **`-2.31% ✅`** | **`-0.26% ✅`** |
|    pool-together | **`+1.47% ❌`** |                |                |
|          uniswap | **`-1.18% ✅`** |                |                |
| yield_liquidator |  **`-0.3% ✅`** | **`-0.34% ✅`** |  **`-0.1% ✅`** |
|         zeppelin | **`+1.66% ❌`** | **`+1.53% ❌`** | **`-0.02% ✅`** |

##### `legacy-optimize-evm+yul`
|          project |  bytecode_size | deployment_gas |     method_gas |
|:----------------:|---------------:|---------------:|---------------:|
|           colony | **`-0.01% ✅`** |                |                |
|        elementfi | **`-0.01% ✅`** |                |                |
|              ens | **`-0.01% ✅`** |          `+0%` |           `0%` |
|            euler | **`-0.06% ✅`** |                |                |
|              gp2 | **`-0.04% ✅`** |                |                |
|  perpetual-pools | **`-0.02% ✅`** | **`-0.02% ✅`** |          `-0%` |
|    pool-together | **`+0.01% ❌`** |                |                |
| yield_liquidator | **`-0.02% ✅`** | **`-0.01% ✅`** |           `0%` |
|         zeppelin | **`-0.01% ✅`** | **`-0.01% ✅`** | **`-0.08% ✅`** |

#### Gas diff stats
| File name                                                                   | IR optimized   | Legacy optimized   | Legacy   |
|-----------------------------------------------------------------------------|----------------|--------------------|----------|
| `externalContracts/ramanujan_pi.sol`                                        | 14%            |                    |          |
| `externalContracts/prbmath_signed.sol`                                      | 1%             |                    |          |
| `array/function_array_cross_calls.sol`                                      | 1%             |                    |          |
| `abiEncoderV2/abi_encode_v2_in_function_inherited_in_v1_contract.sol`       | 1%             | -0%                |          |
| `various/address_code.sol`                                                  | 1%             |                    |          |
| `events/event_emit_from_other_contract.sol`                                 | -1%            |                    |          |
| `functionCall/gas_and_value_brace_syntax.sol`                               | -2%            |                    |          |
| `functionCall/gas_and_value_basic.sol`                                      | -2%            |                    |          |
| `functionCall/external_call_to_nonexisting_debugstrings.sol`                | -2%            |                    |          |
| `inlineAssembly/transient_storage_selfdestruct.sol`                         | -3%            |                    |          |
| `externalContracts/base64.sol`                                              | -3%            |                    |          |
| `functionCall/external_call_to_nonexisting.sol`                             | -4%            |                    |          |
| `byte_array_to_storage_cleanup.sol`                                         | -4%            |                    |          |
| `array/create_memory_array.sol`                                             | -8%            |                    |          |